From d333b9603b2edd4099b7a0c2950b4e1ec3a15923 Mon Sep 17 00:00:00 2001
From: Rob Herring <r.herring@freescale.com>
Date: Mon, 8 Sep 2008 13:23:37 -0500
Subject: [PATCH] ENGR00069686-1 usb device: add mx51 work-arounds

Enable USB-OTG device mode for i.MX51. This includes setting
the PHY to 19.2MHz instead of keeping the default of 24MHz. This
also includes a software workaround for 31-byte HW issue.

Signed-off-by: Dinh Nguyen <Dinh.Nguyen@freescale.com>
---
 arch/arm/plat-mxc/Kconfig          |    2 +-
 arch/arm/plat-mxc/Makefile         |    1 +
 arch/arm/plat-mxc/usb_common.c     |    6 ++++++
 drivers/usb/gadget/arcotg_udc.h    |    4 ++--
 include/asm-arm/arch-mxc/arc_otg.h |   15 +++++++++++++--
 5 files changed, 23 insertions(+), 5 deletions(-)

diff --git a/arch/arm/plat-mxc/Kconfig b/arch/arm/plat-mxc/Kconfig
index ef7720a..8fdee23 100644
--- a/arch/arm/plat-mxc/Kconfig
+++ b/arch/arm/plat-mxc/Kconfig
@@ -95,7 +95,7 @@ config ISP1504_MXC_OTG
 config UTMI_MXC
 	bool
 	default y if USB_EHCI_FSL_UTMI || USB_GADGET_FSL_UTMI
-	depends on ARCH_MX35 || ARCH_MX37
+	depends on ARCH_MX35 || ARCH_MX37 || ARCH_MX51
 
 # set iff we need the 1301 transceiver code
 config ISP1301_MXC
diff --git a/arch/arm/plat-mxc/Makefile b/arch/arm/plat-mxc/Makefile
index 04e4173..85d9fd4 100644
--- a/arch/arm/plat-mxc/Makefile
+++ b/arch/arm/plat-mxc/Makefile
@@ -20,6 +20,7 @@ obj-$(CONFIG_ARCH_MX27) += dma_mx2.o usb_common.o
 obj-$(CONFIG_ARCH_MX3) += dptc.o usb_common.o
 obj-$(CONFIG_ARCH_MX35) += usb_common.o serialxc.o
 obj-$(CONFIG_ARCH_MX37)     += usb_common.o utmixc.o
+obj-$(CONFIG_ARCH_MX51)	+= usb_common.o utmixc.o
 
 # LEDs support
 obj-$(CONFIG_LEDS) += leds.o
diff --git a/arch/arm/plat-mxc/usb_common.c b/arch/arm/plat-mxc/usb_common.c
index 95e5b99..899ed5c 100644
--- a/arch/arm/plat-mxc/usb_common.c
+++ b/arch/arm/plat-mxc/usb_common.c
@@ -540,6 +540,12 @@ static void otg_set_utmi_xcvr(void)
 	/* Enable UTMI interface in PHY control Reg */
 	USB_PHY_CTR_FUNC |= USB_UTMI_PHYCTRL_UTMI_ENABLE;
 
+	if (cpu_is_mx51()) {
+		/* Set the PHY clock to 19.2MHz */
+		USB_PHY_CTR_FUNC2 &= ~USB_UTMI_PHYCTRL2_PLLDIV_MASK;
+		USB_PHY_CTR_FUNC2 |= 0x01;
+	}
+
 	if (UOG_HCSPARAMS & HCSPARAMS_PPC)
 		UOG_PORTSC1 |= PORTSC_PORT_POWER;
 
diff --git a/drivers/usb/gadget/arcotg_udc.h b/drivers/usb/gadget/arcotg_udc.h
index 9048b5f..cc7580f 100644
--- a/drivers/usb/gadget/arcotg_udc.h
+++ b/drivers/usb/gadget/arcotg_udc.h
@@ -24,8 +24,8 @@
 #define FALSE 0
 
 #define MSC_BULK_CB_WRAP_LEN 31
-#define USE_MSC_WR(len) ((cpu_is_mx37_rev(CHIP_REV_1_0) == 1) && \
-	((len) == MSC_BULK_CB_WRAP_LEN))
+#define USE_MSC_WR(len) (((cpu_is_mx37_rev(CHIP_REV_1_0) == 1) ||\
+	(cpu_is_mx51() == 1)) && ((len) == MSC_BULK_CB_WRAP_LEN))
 
 /* Iram patch */
 #ifdef CONFIG_USB_STATIC_IRAM_PPH
diff --git a/include/asm-arm/arch-mxc/arc_otg.h b/include/asm-arm/arch-mxc/arc_otg.h
index 7c7e09d..037d0a2 100644
--- a/include/asm-arm/arch-mxc/arc_otg.h
+++ b/include/asm-arm/arch-mxc/arc_otg.h
@@ -16,7 +16,13 @@
 #define USB_OTGREGS_BASE	(OTG_BASE_ADDR + 0x000)
 #define USB_H1REGS_BASE		(OTG_BASE_ADDR + 0x200)
 #define USB_H2REGS_BASE		(OTG_BASE_ADDR + 0x400)
+#ifdef CONFIG_ARCH_MX51
+#define USB_H3REGS_BASE		(OTG_BASE_ADDR + 0x600)
+#define USB_OTHERREGS_BASE	(OTG_BASE_ADDR + 0x800)
+#else
 #define USB_OTHERREGS_BASE	(OTG_BASE_ADDR + 0x600)
+#endif
+
 
 #define USBOTG_REG32(offset)	(*((volatile u32 *)(IO_ADDRESS(USB_OTGREGS_BASE + (offset)))))
 #define USBOTG_REG16(offset)	(*((volatile u16 *)(IO_ADDRESS(USB_OTGREGS_BASE + (offset)))))
@@ -55,7 +61,7 @@
 #define UOG_BURSTSIZE		USBOTG_REG32(0x160)	/* host ctrlr embedded TT async buf status */
 #define UOG_TXFILLTUNING	USBOTG_REG32(0x164)	/* TX FIFO fill tuning */
 #define UOG_ULPIVIEW		USBOTG_REG32(0x170)	/* ULPI viewport */
-#define	UOG_CFGFLAG		USBOTG_REG32(0x180)	/* configflag (supports HS) */
+#define UOG_CFGFLAG		USBOTG_REG32(0x180)	/* configflag (supports HS) */
 #define UOG_PORTSC1		USBOTG_REG32(0x184)	/* port status and control */
 /* end EHCI registers: */
 #define UOG_OTGSC		USBOTG_REG32(0x1a4)	/* OTG status and control */
@@ -134,7 +140,8 @@
  */
 #define USBCTRL			USBOTHER_REG(0x00)	/* USB Control register */
 #define USB_OTG_MIRROR		USBOTHER_REG(0x04)	/* USB OTG mirror register */
-#define USB_PHY_CTR_FUNC        USBOTHER_REG(0x08)      /* OTG UTMI PHY Function Control register */
+#define USB_PHY_CTR_FUNC	USBOTHER_REG(0x08)      /* OTG UTMI PHY Function Control register */
+#define USB_PHY_CTR_FUNC2	USBOTHER_REG(0x0c)      /* OTG UTMI PHY Function Control register */
 
 /*
  * register bits
@@ -289,6 +296,10 @@
 /* USB_PHY_CTRL_FUNC */
 #define USB_UTMI_PHYCTRL_UTMI_ENABLE   0x01000000
 
+/* USB_PHY_CTRL_FUNC2*/
+#define USB_UTMI_PHYCTRL2_PLLDIV_MASK		0x3
+#define USB_UTMI_PHYCTRL2_PLLDIV_SHIFT		0
+
 /* ULPIVIEW register bits */
 #define ULPIVW_OFF		(0x170)
 #define ULPIVW_WU		(1 << 31)	/* Wakeup */
-- 
1.5.4.4

